src/lib/provable/bigint.ts

```diff
@@ -0,0 +1,898 @@
+  },
+  464: {
+    limb_num: 4,
+    limb_size: 116n,
+    mask: (1n << 116n) - 1n,
+    MAX: (1n << 464n) - 1n,
+  },
+  580: {
+    limb_num: 5,
+    limb_size: 116n,
+    mask: (1n << 116n) - 1n,
+    MAX: (1n << 580n) - 1n,
+  },
+  1044: {
+    limb_num: 9,
+    limb_size: 116n,
+    mask: (1n << 116n) - 1n,
+    MAX: (1n << 1044n) - 1n,
+  },
+  2088: {
+    limb_num: 18,
+    limb_size: 116n,
+    mask: (1n << 116n) - 1n,
+    MAX: (1n << 2088n) - 1n,
+  },
+  4176: {
+    limb_num: 36,
+    limb_size: 116n,
+    mask: (1n << 116n) - 1n,
+    MAX: (1n << 4176n) - 1n,
+  },
+};
+
+/**
+ * Creates a class representing a ProvableBigInt with modular arithmetic capabilities.
+ * This is particularly useful for implementing prime fields that doesn't fit into the native field.
+ *
+ * ```ts
+ * const BigInt521 = createProvableBigInt(2n ** 521n - 1n); // creates a class for 521-bit integers
+ * ```
+ *
+ * `createProvableBigInt(modulus, config?)` takes two parameters:
+ * - `modulus`: The modulus of the field (must be a prime)
+ * - `config`: Optional configuration for custom limb size and numbers
+ *
+ * The returned class supports comprehensive arithmetic operations including:
+ * - Basic operations: addition, double, subtraction, multiplication, square, division
+ * - Advanced operations: inverse, negate, sqrt, power
+ * - Comparison operations: equals, assertEquals, greaterThan, lessthan, greaterThanOrEqual, lessThanOrEqual
+ * - Conversion methods: fromBigInt, toBigInt, fromFields, toFields, fromBits, toBits
+ *
+ * Implementation details:
+ *
+ * Internally, a ProvableBigInt is represented as an array of Field elements (limbs),
+ * where each limb holds 116 bits as default. The total size is determined by the configuration,
+ * with preset options supporting different bit lengths:
+ * - 348 bits (3 limbs)
+ * - 464 bits (4 limbs)
+ * - 580 bits (5 limbs)
+ * - 1044 bits (9 limbs)
+ * - 2088 bits (18 limbs)
+ * - 4176 bits (36 limbs)
+ *
+ * Each arithmetic operation ensures the result is a valid element of the prime field.
+ *
+ * @example
+ * ```ts
+ * // Create a Provable BigInt class with modulus 2^521 - 1
+ * const BigInt521 = createProvableBigInt(2n ** 521n - 1n);
+ *
+ * // Create instances
+ * const a = BigInt521.fromBigInt(123n);
+ * const b = BigInt521.fromBigInt(456n);
+ * const c = BigInt521.fromBigInt(1024n);
+ *
+ * // Perform operations
+ * const sum = a.add(b);
+ * const double = a.add(b);
+ * const diff = a.sub(b);
+ * const product = a.mul(b);
+ * const square = a.square();
+ * const quotient = a.div(b);
+ * const inverse = a.inverse();
+ * const negation = a.negate();
+ * const power = a.pow(b);
+ * const sqrt = c.sqrt();
+ *
+ * ```
+ *
+ * The class automatically handles modular reduction after
+ * arithmetic operations to maintain valid representations. All operations are
+ * designed to be provable and optimised for less constraints.
+ *
+ * @param modulus The modulus for the big integer arithmetic (must be prime)
```

Comment:
I pointed this out earlier, but: either assert it is a prime, or specify that if it is not a prime then some operations are not possible.